Generator.drawTab(), Minecraft.drawFaceTab() and Minecraft.drawFaceTabs()

pixelpapercraft opened the pull request:
This PR introduces three new functions

## Generator.drawTab()

This function takes a `rectangle` and uses it as a **boundary** for drawing a tab.

The `orientation` indicates which edge of the rectangle the tab should be drawn on.

Tabs have optional fold lines.

And the slope of the tab defaults to 60 degrees.

```
let drawTab = (
  rectangle: Builder.rectangle,
  orientation: [#North | #South | #East | #West],
  ~showFoldLine: bool=true,
  ~tabAngle: float=45.0,
  (),
)
```

Example:

```
let rectangle = (10, 10, 100, 50)
Generator.drawTab(rectangle, #North, ())
```

<img width="119" alt="image" src="https://user-images.githubusercontent.com/85809598/174468970-9b4638cd-43c8-45f3-8364-d89ccbd55d90.png">

## Minecraft.drawFaceTab()

Actually three functions here:

* Minecraft.drawFaceTab()
* Minecraft.drawFaceTabs()
* Minecraft.setTabSize()

This simplifies drawing tabs for faces on a cuboid.

Here we provide the `face` of the cuboid, which is a rectangle for the "right", "left", "top", etc.

We indicate if we want a North, South, East, or West tab.

Default size is 24 pixels. This is the maximum "height" of the tab, but the height may be smaller if the tab diagonal lines intersect.

Default angle is 45 degrees.

```
let drawFaceTab = (
  face: Builder.rectangle,
  side: [#North | #South | #East | #West],
  ~size: int=24,
  ~showFoldLine: bool=true,
  (),
)
```

Example:

```
let face = (100, 100, 64, 64)
Generator.drawTexture("Steve", head.right, face, ())
Minecraft.setTabSize(36)
Minecraft.drawFaceTabs(face, [#West, #North, #South], ())
```

pixelpapercraft commented:
Testing page output:

![image](https://user-images.githubusercontent.com/85809598/174468860-5212b310-7985-4ac6-ae12-e3b18bff7cb1.png)

pixelpapercraft commented:
@NinjolasNJM @PaperDoggy could I get your feedback on this PR please.

pixelpapercraft commented on src/generators/Minecraft.res:
```diff
@@ -240,3 +240,236 @@ module MinecraftCharacter = {
+//     +----+----+      ---   ---
+//     p1        p4
+//
+let drawRectTab = (
```
I might rename this to just `drawTab()`.

pixelpapercraft commented:
A few more tests:

![image](https://user-images.githubusercontent.com/85809598/174502849-3b6f7889-6484-4820-8df6-b92ec9340414.png)

![image](https://user-images.githubusercontent.com/85809598/174502852-2c8b9651-14f8-4137-b7b6-3745d2989ce7.png)

NinjolasNJM commented:
My thoughts: 

- I think that it would be more useful if the default angle was 45 degrees instead of 60 degrees. It is used in a little more than half the generators, and seems to be more common in designs than greater angles because it is easier to draw by hand, won't overlap other tabs both when the model is glued together and when drawn perpendicular to each other (see the stair model in the block generator for the latter). When cutting a model with 45 degree tabs, tabs at corners can also be cut out without turning when going from one tab to the next.
- Likewise, I think that a default size should be 32 pixels. While other generators might require lower heights like 8, 12 or 16 pixels, the block generator, which would probably benefit from a tabs function the most as it would dramatically cut down on its amount if images used, would be using 32 pixels as its standard height. Alternatively, the default could be a function of the dimensions of the face- most of the time, tab heights seem to be around 25% of one face's length.
- It fee like with `drawFaceTab()` drawing only one tab but still having only a rectangle for its input, it might get confusing for people to know which to use, so that only one of the two functions is used at all. What I would have instead is a `drawFaceTabs()` that instead of drawing a single tab in a single direction, could have a parameter for each side whose face needed drawing- that way in `drawCuboid()` for instance, the function would be called once for each face for better organization. Also, instead of a rectangle it could eventually use a Face.t type as an input, so that it can be used with rotated faces and the like.

pixelpapercraft commented:
Great feedback @NinjolasNJM 

> I think that it would be more useful if the default angle was 45 degrees

I originally started using 60 degrees because it was easier to fold and stick, but it is more common to see 45 degrees. Let's change it 👍 

> Likewise, I think that a default size should be 32 pixels.

I've introduced a new function `Minecraft.setTabSize()` which is global to all tabs. This is a better strategy. I assigned 24 as the default based on the character model which I am hoping is more representative of the typical model size.

> It fee like with drawFaceTab() drawing only one tab but still having only a rectangle for its input, 

Nice idea, I've added a `drawFaceTabs()`.

PaperDoggy commented:
Looks really great, but I see some problems with this.
1) I don`t really see that much use in DrawFaceTab and DrawFaceTabs just because 90% of all pieces are cuboids(So the DrawCuboidTabs would be more practical) and if there even is some sort of usability for DrawFaceTab and DrawFaceTabs then it is for something like chicken legs or some other things. And here there is also a problem because if we a glueing these 2d things to another part using tabs, those tabs will be open and that will look a bit ugly on the completed papercraft. 
DrawFaceTab and DrawFaceTabs just won`t be useful for 3D stuff just because all 3D pieces that people use are cuboids and even if someone uses something different than it will be very very complicated and then it will be just easier to use just DrawFold
2) I fully agree with Ninjolas about 45*. I`m using those in my papercrafts and they are really good because it`s really important so that parts dont overlap not just on design , but just when you are making the papercraft itself.
3)As I mentioned before the DrawCuboidTabs would be a cool thing to have and as almost all cuboid pieces use the same "Pattern" of placing the tabs I think that there can just be that one pattern and the issue that can happen is when tab in actually longer than the face it will be glued to, I`m not sure if this is a problem with the DrawFold thing because we can chose the size. 

Everything else looks really great.

pixelpapercraft commented:
Great feedback @PaperDoggy

Having a draw cuboid tabs is a good idea. Once @NinjolasNJM's PR for cuboids is merged then we can work on that one.

I believe we can use the other functions for designs such as steps:

<img width="580" alt="image" src="https://user-images.githubusercontent.com/85809598/174894143-cf5a60dc-e2bd-4188-acf7-a24ab705df6a.png">

Also, would you be able to explain this point a bit more?

> and the issue that can happen is when tab in actually longer than the face it will be glued to

Thanks!

PaperDoggy commented:
![Examplsss](https://user-images.githubusercontent.com/88270934/174966842-a5052db8-35e1-444f-929a-de25e54c3507.png)
